src/riot-rs-threads/src/lib.rs

```diff
@@ -172,22 +168,29 @@ impl Threads {
         }
     }
 
-    /// Sets the state of a thread.
+    /// Sets the state of a thread and triggers the scheduler if needed.
     ///
-    /// This function handles adding/ removing the thread to the Runqueue depending
+    /// This function also handles adding/ removing the thread to the Runqueue depending
     /// on its previous or new state.
     ///
     /// # Panics
     ///
     /// Panics if `pid` is >= [`THREADS_NUMOF`].
     fn set_state(&mut self, pid: ThreadId, state: ThreadState) -> ThreadState {
-        let thread = &mut self.threads[usize::from(pid)];
-        let old_state = thread.state;
-        thread.state = state;
-        if old_state != ThreadState::Running && state == ThreadState::Running {
-            self.runqueue.add(thread.pid, thread.prio);
-        } else if old_state == ThreadState::Running && state != ThreadState::Running {
-            self.runqueue.pop_head(thread.pid, thread.prio);
+        let thread = self.get_unchecked_mut(pid);
+        let old_state = core::mem::replace(&mut thread.state, state);
+        let prio = thread.prio;
+
+        if state == ThreadState::Running {
+            self.runqueue.add(pid, prio);
+            self.schedule_if_higher_prio(pid, prio);
+        } else if old_state == ThreadState::Running {
+            // A running thread is only set into a non-running state
+            // if it itself initiated it.
+            debug_assert_eq!(Some(pid), self.current_pid());
+
+            self.runqueue.pop_head(pid, prio);
+            schedule();
```

Comment:
I removed the checks that the respective other state is `!= ThreadState::Running`, because we never have that case in our code.

Interestingly, if I don't remove them `bench_sched_yield` only takes 574 instead of 580 ticks for the nRF52840, which is very weird given that the benchmark doesn't even use this method apart from the initial thread creation. Any idea why that could be @kaspar030?
(`bench_sched_flags` performs as expected better without the checks).

Reply:
> I removed the checks that the respective other state is `!= ThreadState::Running`, because we never have that case in our code.

Hm, while that may well be, are we ok if we made a mistake here? E.g., in the first case, `runqueue.add()` calls `runqueue.queues.push()`, which with a quick glance probably won't like if a thread is added a second time.

Reply:
`CList::push` prevents that thread is added twice, by early-returning if `self.next_idxs[n as usize] != Self::sentinel()`.

And in the second `else if` branch the check `state != ThreadState::Running` then is redundant because that's already catched by the first branch.

Reply:
no, fine, just verifying. :slightly_smiling_face: